src/utils/UVGenerator.js

```diff
@@ -0,0 +1,173 @@
+// https://github.com/mozilla/Spoke/commit/9701d647020e09d584885bd457eb225e9995c12f
+import XAtlas from '@gfodor/xatlas-web';
+import { Float32BufferAttribute, Uint32BufferAttribute } from 'three';
+
+const AddMeshStatus = {
+	Success: 0,
+	Error: 1,
+	IndexOutOfRange: 2,
+	InvalidIndexCount: 3
+};
+
+export class UVGenerator {
+
+	constructor() {
+
+		this._module = null;
+		this.channel = 1;
+
+	}
+
+	async init() {
+
+		if ( this._module ) return;
+
+		const wasmurl = new URL( '../../node_modules/@gfodor/xatlas-web/dist/xatlas-web.wasm', import.meta.url );
+		this._module = XAtlas( {
+
+			locateFile( path ) {
+
+				if ( path.endsWith( '.wasm' ) ) {
+
+					return wasmurl.toString();
+
+				}
+
+				return path;
+
+			}
+
+		} );
+
+		this._xatlas = await this._module;
+
+	}
+
+	generate( geometries, onProgress = null ) {
+
+		if ( geometries.constructor !== Array ) {
+
+			geometries = [ geometries ];
+
+		}
+
+		const xatlas = this._xatlas;
+		xatlas.createAtlas( );
+
+		const meshInfos = [];
+
+		for ( let i = 0; i < geometries.length; i ++ ) {
+
+			const geometry = geometries[ i ];
+
+			const originalVertexCount = geometry.attributes.position.count;
+			const originalIndexCount = geometry.index.count;
+
+			const meshInfo = xatlas.createMesh( originalVertexCount, originalIndexCount, true, true );
+			xatlas.HEAPU32.set( geometry.index.array, meshInfo.indexOffset / Uint32Array.BYTES_PER_ELEMENT );
+			xatlas.HEAPF32.set( geometry.attributes.position.array, meshInfo.positionOffset / Float32Array.BYTES_PER_ELEMENT );
+			xatlas.HEAPF32.set( geometry.attributes.normal.array, meshInfo.normalOffset / Float32Array.BYTES_PER_ELEMENT );
+
+			if ( geometry.attributes.uv ) {
+
+				xatlas.HEAPF32.set( geometry.attributes.uv.array, meshInfo.uvOffset / Float32Array.BYTES_PER_ELEMENT );
```

Comment:
Why do we need to pass the UVs and normals into the library? I would have expected only position to be needed.

Reply:
I'm not sure if it's strictly necessary - XAtlas has a mode you can run it in where you hand it UVs for each mesh and it will generate a new composed atlas using those UVs. My guess is that these aren't accessed by XAtlas for the case we are using it for.